### ISSUE TITLE
salt-ssh trying to write to wrong log file path

### ISSUE THREAD

**Issue by anlutro** [saltstack/salt#27491]
I use salt-ssh with a Saltfile to change the location of most files. I ran a test with the current **develop** branch and this warning popped up. I can run 2015.8 without problems.

```
$ salt-ssh vagrant-test test.ping
[WARNING ] Failed to open log file, do you have permission to write to /path/to/salt/var/log/salt/master?
```

Relevant bits from my Saltfile:

```
salt-ssh:
  config_dir: ./config
  log_file: ./logs/ssh
```

Master config sets the root directory to the CWD:

```
root_dir: ./
```

**Comment by jfindlay** [comment]
@anlutro, thanks for the report.  This was fixed recently, but I cannot find the pull request, so I am wondering if a merge forward from 2015.8 will fix this for develop.

ping @basepi

**Comment by anlutro** [comment]
OK. If you ping me when a merge forward happens, I'd be happy to test again.

**Comment by jfindlay** [comment]
@anlutro, once #27531 is merged, you should be ready to test again, thanks.

**Comment by basepi** [comment]
It's merged now. The only fixes not in develop are the fixes from 2015.5 included here: https://github.com/saltstack/salt/pull/27525

They'll go in next merge forward.

**Comment by anlutro** [comment]
The issue persists.

**Comment by jfindlay** [comment]
@anlutro, I found the issue: #27063.  This may be a duplicate of that.

**Comment by anlutro** [comment]
I don't think it's related - that issue relates to the `/var/log/salt` directory not being created when a yum package is being installed. Unrelated to salt-ssh and a different log path altogether.

I think the issue here is that salt-ssh shouldn't be trying to log to a `master` file in the first place. salt-ssh isn't a master and shouldn't be logging there, salt-ssh already has its dedicated log file as defined in the Saltfile.

**Comment by jfindlay** [comment]
@anlutro, that makes sense.  I'm juggling too many issues right now.

**Comment by anlutro** [comment]
It turns out, if I provide a `--log-file=whatever` argument to salt-ssh, the log path from my Saltfile is read correctly. Amusingly, the actual argument I pass to `--log-file` doesn't matter, unless I remove the `log_file` option from my Saltfile.

I tried debugging further, but the various mixins in parsers.py are not easy on the brain.

**Comment by basepi** [comment]
That is....super weird. But good to know!

**Comment by anlutro** [comment]
Bumping because this is still an issue in 2016.3 HEAD.

**Comment by anlutro** [comment]
Progress! From 2015.8, some lines in `parsers.py` have changed from this:

    elif value is not None and value != default:

to this:

    elif value is not None and getattr(option, "explicit", False):

I don't know the reason for this, I'm guessing they're good, but if I revert this change, the warning disappears and logging works as expected.

**Comment by anlutro** [comment]
```
-> if option.dest not in self.config:
(Pdb) print(option)
--log-file
(Pdb) print(option.explicit)
*** AttributeError: 'CustomOption' object has no attribute 'explicit'
```

It looks like optparse never calls `take_action`, which means the `explicit` property never gets set.

I found the related PR: https://github.com/saltstack/salt/pull/25785

@htch Do you have any ideas what the best way to fix this would be?

**Comment by anlutro** [comment]
This seems to have had a regression at some point - with 2016.11 I'm getting similar errors.

```
Failed to create path "/var/log/salt/ssh" - [Errno 13] Permission denied: '/var/log/salt'
```

**Comment by anlutro** [comment]
Actually seeing the same problem with a git checkout of 2016.3.

**Comment by dejadejoder** [comment]
Same pb for me (salt-ssh 2016.11.1 (Carbon)):
`[WARNING ] Failed to open log file, do you have permission to write to /var/log/salt/ssh?`

**Comment by dmurphy18** [comment]
@anlutro Could you open a new issue referencing this issue, and mark it as a regression. This will allow us to track the issue better, given the large number of codes changes since 2015.5.

**Comment by dmurphy18** [comment]
@anlutro Never mind, I had stale browser data with it marked closed. I see that you re-opened, which is just as good

**Comment by ashokrajar** [comment]
ssh_log_file: /tmp/salt-ssh-log.log option in the Saltfile works for me.

ssh_log_file instead of log_file

**Comment by kojiromike** [comment]
Confirmed: setting `ssh_log_file` explicitly does resolve this issue.

**Comment by pkoch** [comment]
+1. Exactly the same here.

**Comment by thatch45** [comment]
I am going to close this out, for 2 reasons:
1. We are trying to migrate work to heist
2. This issue is related to running salt as a non-root user where we have a number of issues that are path related. Those issues should be resolved which would encompass this
